torch/csrc/dynamo/eval_frame.c

```diff
@@ -745,17 +912,27 @@ static PyObject* _custom_eval_frame(
     return NULL;
   } else if (result != Py_None) {
     DEBUG_TRACE("create cache %s", get_frame_name(frame));
-    extra = create_cache_entry(extra, result);
+    CacheEntry* new_cache_entry = create_cache_entry(cache_entry, result);
     Py_DECREF(result);
-    set_cache_entry(frame->f_code, extra);
+    if (extra == NULL) {
+      // Create a new extra state as this is the first time we see this code object.
+      create_and_set_extra_state(frame->f_code, new_cache_entry, frame_state);
+    } else {
+      // Update the existing cache_entry on the extra scratch space.
+      override_cache_entry_on_extra(extra, new_cache_entry);
```

Comment:
Yeah this has the same problem as the other call site. There's one more problem, which is you're using extra == NULL to demarcate whether or not there was already a cache entry on extra, but that's actually an extra invariant. Couldn't you imagine a case where you have no cache entries, but you DO have a frame state? It seems better to just get the truth from the horse's mouth.

Reply:
Thanks for laying it out for me. Yeah, I inlined the code. There is still violation but it is contained in a local region, so hopefully its acceptable?